moshi/src/main/java/com/squareup/moshi/JsonAdapter.java

```diff
@@ -37,7 +37,12 @@
   }
 
   @CheckReturnValue public final @Nullable T fromJson(String string) throws IOException {
-    return fromJson(new Buffer().writeUtf8(string));
+    JsonReader reader = JsonReader.of(new Buffer().writeUtf8(string));
+    T result = fromJson(reader);
+    if (reader.peek() != JsonReader.Token.END_DOCUMENT) {
+      throw new JsonDataException("JSON document was not fully consumed.");
```

Comment:
I would include the next token as additional hint

Reply:
Sounds good, as long as it's consistent with the check in Retrofit's converter.
@swankjesse Any objections?